### ISSUE TITLE
Unclear on How to Start a Custom Kernel on Kubernetes

### ISSUE THREAD

**Issue by bashterm** [jupyter-incubator/enterprise_gateway#437]
I'm trying to set up a custom python3 kernel on kubernetes, and I'm having some difficulty understanding the way in which I'm supposed to do that.

I'm not using spark, and I've looked at the [default Dockerfile](../blob/master/etc/docker/kernel-py/Dockerfile) for an ipython kernel in Kubernetes, however, I'm not clear on if I need to use the spark driver, or what. How exactly does the architecture work, for deploying on vanilla kubernetes?

Thanks for any help

**Comment by lresende** [comment]
You could start with our [documentation](https://jupyter-enterprise-gateway.readthedocs.io/en/latest/getting-started-kubernetes.html) and provide feedback on what is not clear there. 

In summary, you will need:
```
kubectl apply -f https://raw.githubusercontent.com/jupyter-incubator/enterprise_gateway/master/etc/kubernetes/enterprise-gateway.yaml
```
You might also want to pull the docker images to 
```
    docker pull elyra/kubernetes-enterprise-gateway:dev
    docker pull elyra/kubernetes-kernel-py:dev
    docker pull elyra/kubernetes-kernel-tf-py:dev
    docker pull elyra/kubernetes-kernel-r:dev
    docker pull elyra/kubernetes-kernel-scala:dev
```

Note: temporarily, you will need to edit the deployed pod and fix the image name as we are in the process of pushing new/renamed images. You will need to add `kubernetes-` to the existing image name. You could also do this by downloading the yaml file and updating that before issuing a `kubectl apply`.

**Comment by bashterm** [comment]
I have a working deployment of the enterprise gateway itself. I'm wondering how i create a custom kernel to deploy on the cluster through the enterprise gateway. 

It looks to me as if the [spark-exec](../blob/master/etc/docker/kernel-py/spark-exec.sh) is what actually runs the kernel, but I'm able to run the kernel image fine on a kubernetes cluster without spark installed. How is that execution working? Is there something going on in the upstream docker image for the kubespark executor that I can't see?

**Comment by lresende** [comment]
And for customizing images, you can take a look at the [kernelspecs for python & tensorflow](https://github.com/jupyter-incubator/enterprise_gateway/tree/master/etc/kernelspecs/python_tf_kubernetes). In that, we set the custom image 
```
 "process_proxy": {
    "class_name": "enterprise_gateway.services.processproxies.k8s.KubernetesProcessProxy",
    "config": {
      "image_name": "elyra/kernel-tf-py:dev"
    }
```
And then, [these](https://github.com/jupyter-incubator/enterprise_gateway/tree/master/etc/kernel-launchers/kubernetes/scripts) are going to be used to launch the pods.

Currently, if you customize, you would want to build a new image and redeploy it to kubernetes

```
make clean build enterprise-gateway
```

The command above will generate a kernelspec archive by merging the files from kernelspecs and scripts appropriately and then build an enterprise-gateway docker image with the customized kernels.

The other approach is to edit the enterprise gateway yaml and enable pv and store the kernelspecs there, where you can start creating+customizing your kernelspecs. 

Now, if what you want to modify is some existing parameters, you can define KERNEL_XXX env variables on the NB2KG image/host and those will be properly propagated to the gateway->kernel.

**Comment by bashterm** [comment]
OK. I've enabled the PV and can upload kernelspecs to it. When I'm making a new kernel image, is there anything special I need to include that isn't in the [Dockerfile](../blob/master/etc/docker/kernel-py/Dockerfile) in this repo? Can I base that image on a raw alpine linux image, or do I need to use the kubespark base image?

**Comment by lresende** [comment]
Not really. We used kubespark in some images to kind reduce the number of images we have to maintain (vanilla versus spark). You can see that, for instance, tensorflow is inheriting from  `tensorflow/tensorflow:1.9.0-py3`. If you base your image on alpine linux, then you will need to make sure that you add all required dependencies to launch your kernel to the image (e.g. if python based, IPython, etc).

**Comment by kevin-bates** [comment]
The key to integrating with EG is having the kernel wrapper in place that handles creation of the ZMQ ports and their conveyance back to EG via the response-address parameter.  this wrapper resides in the scripts directory of the kernelspec and is invoked by the bootstrap-kernel.sh file (or equivalent).

Although we install the complete kernelspecs tar file in each of our kernel images, we really only need the respective sub-tree relative to the image's target kernel.

Also, we have no dependency on the user we create (eg-kernel in this case).  That's merely an example.

**Comment by bashterm** [comment]
OK. So I need to create two kernelspecs then, one that the EG sees in the PV, that points to the correct image, and one that the image uses to start the kernel?

**Comment by kevin-bates** [comment]
Actually, all that is needed in the kernel image is the launcher that wraps the kernel and any other ancillary information - which is typically most easily provided by the `run.sh` but is not necessary.  All of which is invoked by the image's CMD entry.

The key `kernel.json` would be the one in the PV that is used by EG.  However, you may have noticed that this typically just calls a `launch_kubernetes.py` script - along with the various parameters EG needs - which is responsible for creating the kernel pod via the kernel-pod.yaml located in the same scripts folder as the `launch_kubernetes.py`.  At this point, all kernel images share the same kernel-pod.yaml, but should you want to add some custom items, just place yours in the scripts folder.   Note that if even if you don't need custom support when creating the pod, you'll want to have the default `kernel-pod.yaml` in the scripts folder - along with `launch_kubenetes.py`.

I hope that helps.  One of the areas of our docs that isn't in place is Bring-Your-Own-Whatever kind of things - simply because we're fairly new - so this is an extremely helpful exercise for us.  Thank you for patience.

**Comment by bashterm** [comment]
No problem. Since I'm launching an IPython Python3 kernel, can I use the existing launch_ipykernel.py script? Looking through it it doesn't look like anything is specific to python2. Writing a run.sh script also looks like it shouldn't be crazy hard.

Thanks for all the support. This project is super cool!

**Comment by kevin-bates** [comment]
Yes - launch_ipykernel.py should be fine.  Also, you only need to typically build run.sh if you're going to perform something like spark-submit, but if you're just trying to launch the python3 kernel in its own pod, then just follow the model of etc/docker/kernel-tf-py/{Dockerfile,bootstrap-kernel.sh} and etc/kernelspecs/python_tf_kubernetes/kernel.json.

**Comment by bashterm** [comment]
OK. Just for my own understanding, where does the kernel process actually get launched within the docker image? Is it from the launch_ipykernel.py script, where `embed_kernel` is called?

**Comment by kevin-bates** [comment]
yes

**Comment by bashterm** [comment]
OK. How does it know which kernel to start, if there are multiple installed? Is it just running the default one? I don't see the kernel name passed in anywhere.

**Comment by kevin-bates** [comment]
The target kernel is passed via the `name` entry in the json payload of the POST request against `/api/kernels`.  This value essentially references the name of the corresponding kernelspec - which equates to the directory on the notebook/gateway server that contains the kernel.json (and supporting files).  If the json request does not specify a `name` entry, then the configured default kernelspec will be used.

When the gateway starts, the KernelSpecManager (in our case - `RemoteKernelSpecManager` loads the set of kernelspecs (each represented  by `RemoteKernelSpec` instances).  When a start request is received, the referenced `name` is then located in the set of `RemoteKernelSpec` instances.  We overload the KernelSpec classes to handle the `process_proxy` stanza so we can "plugin" our lifecycle manager relative to the underlying resource manager.

**Comment by bashterm** [comment]
Ahh ok. Thanks. That makes sense.
